optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/tri/DroolsAbstractTriConstraintStream.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.tri;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.drools.model.Declaration;
+import org.drools.model.PatternDSL;
+import org.optaplanner.core.api.function.ToIntTriFunction;
+import org.optaplanner.core.api.function.ToLongTriFunction;
+import org.optaplanner.core.api.function.TriFunction;
+import org.optaplanner.core.api.function.TriPredicate;
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.tri.TriConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraint;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraintFactory;
+import org.optaplanner.core.impl.score.stream.drools.bi.DroolsAbstractBiConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsAbstractConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.uni.DroolsAbstractUniConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.uni.DroolsFromUniConstraintStream;
+import org.optaplanner.core.impl.score.stream.tri.InnerTriConstraintStream;
+
+public abstract class DroolsAbstractTriConstraintStream<Solution_, A, B, C>
+        extends DroolsAbstractConstraintStream<Solution_> implements InnerTriConstraintStream<A, B, C> {
+
+    protected final DroolsAbstractTriConstraintStream<Solution_, A, B, C> parent;
+
+    public DroolsAbstractTriConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractTriConstraintStream<Solution_, A, B, C> parent) {
+        super(constraintFactory);
+        if (parent == null && !(this instanceof DroolsJoinTriConstraintStream)) {
+            throw new IllegalArgumentException("Parent of stream (" + this + ") must not be null, or not join stream.");
```

Comment:
typo: "not a join stream"
Actually, this sentence is incorrect: "Parent ... must not be null, or not join stream." implies "parent is not a join stream", while it should be "this is not a join stream"

I'd split this up into 2 separate if checks.